R/tokens_group.R

```diff
@@ -22,24 +24,26 @@
 #' # with fill
 #' tokens_group(toks, groups = factor(c(1, 1, 2, 2), levels = 1:3))
 #' tokens_group(toks, groups = factor(c(1, 1, 2, 2), levels = 1:3), fill = TRUE)
-tokens_group <- function(x, groups = docid(x), fill = FALSE) {
+tokens_group <- function(x, groups = docid(x), fill = FALSE, env = NULL) {
     UseMethod("tokens_group")
 }
 
 #' @export
-tokens_group.default <- function(x, groups = docid(x), fill = FALSE) {
+tokens_group.default <- function(x, groups = docid(x), fill = FALSE, env = NULL) {
     check_class(class(x), "tokens_group")
 }
 
 #' @export
-tokens_group.tokens <- function(x, groups = docid(x), fill = FALSE) {
-    x <- as.tokens(x)
+tokens_group.tokens_xptr <- function(x, groups = docid(x), fill = FALSE, env = NULL) {
+    
+    if (is.null(env))
+        env <- parent.frame()
     if (missing(groups)) {
         field <- NULL
         groups <- docid(x)
     } else {
         field <- deparse(substitute(groups))
-        groups <- eval(substitute(groups), get_docvars(x, user = TRUE, system = TRUE), parent.frame())
+        groups <- eval(substitute(groups), get_docvars(x, user = TRUE, system = TRUE), env)
```

Comment:
Under what circumstances would users want to specify `env`?

Reply:
I needed to pass `env` from `tokens_group.tokens()` to `tokens_group.tokens_xptr()` because the global environment is not the parent frame for `tokens_group.tokens_xptr()` (it is grand parent frame). I am no expert on NSE so you are welcome to suggest a better approach.

More broadly, it is hard to use `tokens_group()` in custom functions if `eval()` searches only the parent frame. `env` is what Wickham calls "[escape hatch](http://adv-r.had.co.nz/Computing-on-the-language.html)".

Reply:
That's exactly right, and our current approach also provides that escape hatch. 

But I also need to study this more. I just ran a bunch of tests and while I thought that if not found in the docvars, it would search the calling environment, that does not appear to be the case. It's also not how the tidyverse method in **dplyr** works either.

``` r
myfun1 <- function(x) {
    logvar <- sample(c(TRUE, FALSE), 
                     size = nrow(x),
                     replace = TRUE)
    myfun2(x)
}

myfun2 <- function(y) {
    dplyr::filter(y, logvar)
}

# fails
myfun1(iris)
#> Error in `dplyr::filter()`:
#> ℹ In argument: `logvar`.
#> Caused by error:
#> ! object 'logvar' not found
#> Backtrace:
#>      ▆
#>   1. ├─global myfun1(iris)
#>   2. │ └─global myfun2(x)
#>   3. │   ├─dplyr::filter(y, logvar)
#>   4. │   └─dplyr:::filter.data.frame(y, logvar)
#>   5. │     └─dplyr:::filter_rows(.data, dots, by)
#>   6. │       └─dplyr:::filter_eval(...)
#>   7. │         ├─base::withCallingHandlers(...)
#>   8. │         └─mask$eval_all_filter(dots, env_filter)
#>   9. │           └─dplyr (local) eval()
#>  10. └─base::.handleSimpleError(...)
#>  11.   └─dplyr (local) h(simpleError(msg, call))
#>  12.     └─rlang::abort(message, class = error_class, parent = parent, call = error_call)

# works
logvar <- sample(c(TRUE, FALSE), 
                 size = nrow(iris),
                 replace = TRUE)
myfun1(iris) |>
    tibble::tibble()
#> # A tibble: 79 × 5
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>  
#>  1          5.1         3.5          1.4         0.2 setosa 
#>  2          4.9         3            1.4         0.2 setosa 
#>  3          4.7         3.2          1.3         0.2 setosa 
#>  4          5.4         3.9          1.7         0.4 setosa 
#>  5          4.6         3.4          1.4         0.3 setosa 
#>  6          5           3.4          1.5         0.2 setosa 
#>  7          4.4         2.9          1.4         0.2 setosa 
#>  8          4.9         3.1          1.5         0.1 setosa 
#>  9          4.8         3            1.4         0.1 setosa 
#> 10          4.3         3            1.1         0.1 setosa 
#> # ℹ 69 more rows
```

<sup>Created on 2023-04-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
I will leave this issue with you.